### ISSUE TITLE
Distributed Training

### ISSUE THREAD

**Issue by jinliangwei** [denizyuret/Knet.jl#303]
Hi Knet developers,

First of all, thank you developing a native neural network engine for Julia. I sincerely appreciate your effort.

I have been working on building distributed systems for machine learning. The major focus for me now is building this Orion framework (https://github.com/jinliangwei/orion, which I just open sourced today) for automatically parallelizing for loops and distributing the computation across a cluster of machines. Iterations of that for loop may have random reads and writes to a global address space (many multiple dimension arrays); the parallelization makes sure the dependences between iteration are properly maintained. Please see our paper at SysML'18 if you'd like more details (http://www.sysml.cc/doc/172.pdf).

As I am adding applications/benchmarks to Orion, I found Knet might be a great use case for Orion. With modest effort, we can have data-parallel training across multiple GPU machines and possibly model-parallel too.

Many existing DL frameworks (TF, MxNet, etc) support distributed training. But they are not so easy to use, especially if you want to implement a network from scratch for distributed training. There is little information about how to do it properly and to the best of my knowledge, few people use that, partially for that reason.

Would you be interested in distributed training for Knet?

In order to use Orion to do data-parallel training for Knet, there are a few things in the Knet application program need to change:

1. Your model parameters need to be created as one or multiple Orion distributed arrays (i.e. DistArray). This is a multi-dimensional array and it's element can be any Julia type. So for a 10-layer network, you can create a 1d DistArray of length 10 where each element is a layer. The storage of those DistArrays are partitioned across a set of server machines by Orion.

2. You need to create a DistArrayBuffer to hold gradients. This essentially is a buffer in each worker machine for accumulating the gradients. Orion is responsible to send to gradients to server and apply them to parameters.

3. Your dataset needs to be a DistArray as well, where each element is a minibatch of data samples.

4. One restriction of Orion is that all these DistArrays need to be defined as global variables, before your training loop that iterates over all data samples.

5. Instead of using update! to apply gradients to weights, you write the gradients to the DistArrayBuffer from 2. To have the servers properly apply gradients with different optimization methods (adagrad, momentum, etc), you need to define an element-wise (layer-wise) function and "share" it with all servers. You many want to have a look at this sparse LR example for how to apply buffered gradients in Orion: https://github.com/jinliangwei/orion/blob/dev_0.6/apps/lr/orion_lr_adarevision.jl#L94

Does the proposal above make sense to you, especially as I see discussions on changing Knet's interface and programming style?

Many Knet examples are interference with pre-trained models. The training examples are mostly NLP models. Are there examples/benchmarks for training more complex models, like resnet? Do you foresee any problems for implementing them?
